### PR TITLE
[POC] [Form] Support get/set accessors for form fields

### LINKED ISSUE #37597: [RFC] [Form] Improve handling of domain objects in forms
# Description

This issue originated as a discussion thread in the Symfony Slack. To preserve the conversation and get more input from maintainers and the community, I'm creating this RFC.

People have been arguing for working with domain transfer objects (DTOs) in the form documentation, as there are a few problem to directly bind the form to a managed object (e.g. a Doctrine entity):
* Data is written to the entity without checking for validity first
* Subsequent validation may flag the form as invalid, but doesn't restore original data
* Subsequently flushing the managed object in the manager (e.g. EntityManager when using Doctrine) may write invalid data to the database and potentially affect unrelated managed objects
* The PropertyAccess only supports accessing properties through accessor methods that are rather hardcoded.
* Using separate methods for write and read operations (e.g. `Category::getName()` and `Category::rename($name)` is not possible

To solve this, users can create a DTO to hold the data and apply it to the managed object after it has been validated. However, this potentially adds a bunch of boilerplate code (e.g. a class with public properties and a `apply` method to apply changed data to the form). In many instances, this can be avoided.

There is a bundle that adds `read_property_path` and `write_property_path` to the form component ([Example](https://github.com/sensiolabs-de/rich-model-forms-bundle/blob/main/docs/mapping.md)). However, @linaori was quick to point out that using strings can cause issues with IDE integration, e.g. by a refactoring not catching the usage in a form as it's not a code reference. @javiereguiluz then suggested the following:

# Example

```php
$builder->add('name', TextType::class, [
    'get' => function (?Category $category): ?string { return $category !== null ? $category->getName() : null; },
    'set' => function (Category $category, $value): void { $category->rename($value); },
]);
```

Note that the getter in this case could be omitted since it replicates the default behaviour.

With short closures, the example gets even more concise:
```php
$builder->add('name', TextType::class, [
    'get' => fn (?Category $category): ?string =>  $category !== null ? $category->getName() : null,
    'set' => fn (Category $category, $value): void => $category->rename($value),
]);
```

This does not solve the problem of validation only being run after modifying the object, but it gives people more freedom when designing the public API for their domain objects, and may avoid the need for a DTO in some instances.

# Future scope

## Exception handling

Exceptions while calling any of the closures will lead to a 500 error. This could be avoided by expanding the closure to catch the exception, but then the question is how this can be communicated to the user. The bundle mentioned above introduces a `expected_exception` form option. If an exception of the given type is thrown, the exception is caught and it is added as a form error. While this may offer more flexibility, I think it introduces yet another form option that people have to understand. An alternative would be to always catch a `ValidationExceptionInterface` interface provided by the form component. The `ValidationException` class would provide a generic exception for people to use. This allows doing the following:

* Users can implement the interface in their domain exceptions:
```php
// Exception class
class NameNotAllowedException extends \RuntimeException implements ValidationExceptionInterface
{
    // ...
}

// Domain object
class Category
{
    public function rename(string $name): void
    {
        if (...) {
            throw NameNotAllowedException::notAllowed($name);
        }

        ...
    }
}
```
* Alternatively, users can handle a specific exception in the closure:
```php
$builder->add('name', TextType::class, [
    'set' => function (Category $category, $value): void {
        try {
            $category->rename($value);
        } catch (DomainException $e) {
            throw ValidationException::fromPrevious($e);
        }
    },
]);
```
* The functionality above could also be provided by a closure:
```php
// DomainValidationHandler.php
public function wrap(string $exceptionClass, Closure $closure): Closure
{
    $code = <<<'CODE'
        return function (object $object, $value): void {
            try {
                $closure($object, $value);
            } catch (%s $e) {
                throw ValidationException::fromPrevious($e);
            }
        };
        CODE;

    return eval(sprintf($code, $exceptionClass));
}

// In form
$builder->add('name', TextType::class, [
    'set' => DomainValidationHandler::wrap(
        DomainException::class,
        fn (Category $category, $value): void => $category->rename($value)
    ),
]);
```

## Change form validation rules

When binding a form, form data is sent through the data transformers, then applied to the object, and last but not least validated by the validation component. This reads constraints from the form as well as from the object and validates all constraints. It may be sensible to change the order when binding:
* Take data from form and run through data transformers. Store transformed data
* Run constraints defined on form fields on model data. Any validation errors are added to the form. If errors are found, binding the form is aborted without changing the object (or creating one if none was passed)
* Apply data to the object. This ensures that any early constraints (e.g. invalid types in the form) can be caught beforehand. Exception handling as described above is optional here.
* Run validation constraints defined on the object. Again, errors are added to the form.

The order above would allow people to keep form-specific validation logic in the form (which would otherwise be put in a DTO) while doing other, potentially more complex validation after data is applied to the object:
```php
$builder->add('name', TextType::class, [
    'set' => DomainValidationHandler::wrap(
        DomainException::class,
        fn (Category $category, $value): void => $category->rename($value)
    ),
    'constraints' => [
        new UniqueEntity([
            'entityClass' => Category::class,
            'fields' => ['name'],
            'id' => // ID of entity being edited will have to be passed from options
    ],
]);
```

# Next steps

For now, I'm looking for feedback from people that have previously created DTOs for their form <-> entity mappings, as well as from users using the bundle mentioned above. I've used the bundle myself, but my last work with Symfony Forms was almost 2 years ago, so I'm a bit outdated when it comes to form. After that, I could offer to build a prototype for the functionality outlined above.

### PR DESCRIPTION
**Pull request by alcaeus**
| Q             | A
| ------------- | ---
| Branch?       | 5.2
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Tickets       | Fix #37597 
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

This is a proof of concept PR based on the original idea in the RFC above, but already incorporating feedback from the community.

# Brief overview

This adds a new form type extension that defines 'get' and 'set' options for forms. If provided, the extension wraps the existing data mapper (usually a PropertyPathMapper) in an AccessorMapper with the provided accessors. If an accessor is left undefined, this falls back to the wrapped data mapper for compatibility.

# Why?

To explain why we do this, let's consider the following gist: https://gist.github.com/alcaeus/0b66968ee505ab24bdb71fe7c88325a7

The `Category` entity has a non-standard accessor to rename the category. While this is expressive, it does not play nice with the form component out of the box. To remedy this problem, one could use the RichModelFormsBundle, which allows for changing the write access property path separate from the read property path.
Alternatively, one can create a `EditCategoryProxy` DTO, which maps a `setName` call (that the PropertyAccess component will seek out) to the `rename` method in the `Category` entity.
A more complex solution uses the `EditCategoryTwoStep` DTO, which contains public properties for all fields, then maps them to the `Category` entity in a special `apply` method which needs to called after the form has been validated. The advantage of this is that it does not leave an entity in an invalid state, since constraints on the DTO are evaluated before `apply` is called (typically by the user before flushing the entity to the manager). It also reduces boilerplate code by doing away with setters and getters and using public properties (since it's just a data holder).

Both DTOs have a significant problem in my view: they are tightly coupled to a specific form, but are separate classes that contain lots of usual boilerplate code. This is where the addition of accessor closures comes into play.

# Basic usage

Forms where you don't specify an accessor (all current forms) behave exactly like before, with no code being changed. They still use a PropertyPathMapper or whatever the user provided. However, upon specifying either 'get' or 'set' option causes cool things to happen. For all examples, please consider the following entity:
```php
final class Category
{
  private string $name;

  public static function fromName(string $name): self
  {
    return new self($name);
  }

  private function __construct(string $name)
  {
    $this->name = $name;
  }

  public function getName(): string
  {
    return $this->name;
  }

  public function rename(string $name): void
  {
    $this->name = $name;
  }
}
```

This class has no typical `setName` accessor as the property access component would expect, so using this entity in a form was previously coupled to using a DTO or a custom data mapper **on the embedding form** to handle this API. With form accessors, the form is defined like this:

```php
$builder->add('name', null, ['set' => function(Category $category, string $name) { $category->rename($value); }]);
```

In this case, getting the value is handled through the PropertyPathMapper, while the value is changed using our closure.

# Advanced usage
Apart from this basic usage, the mapper contains a few features that are nice for people who are more familiar with complex and rich models.

## Using immutable objects

A typical example is storing dates, as it's recommended to always handle `DateTimeImmutable` objects. These return a new instance of themselves with the value applied. The old object is discarded or can be kept for auditing purposes. Let's change the category we had above to being immutable:
```php
final class Category
{
  private string $name;

  public static function fromName(string $name): self
  {
    return new self($name);
  }

  private function __construct(string $name)
  {
    $this->name = $name;
  }

  public function getName(): string
  {
    return $this->name;
  }

  public function rename(string $name): self
  {
    return self::fromName($name);
  }
}
```

Again, the form framework does not support this. With accessors, this becomes increasingly simple:

```php
$builder->add('name', null, ['set' => function(Category $category, string $name) { return $category->rename($value); }]);
```

If the setter returns a non-null value, the mapper behaves as follows:
* If the original data is a scalar value, a scalar value replaces the form data. Otherwise, it is discarded.
* If the original data is an array, returning another array also replaces the form data. Otherwise, it is discarded.
* If the original data is an object, and the returned object is an instance of the same class, this replaces form data. Otherwise, it is discarded.

The form data is overwritten after each step, so any subsequent calls to accessors already operate on the new object.

## Basic validation

Any `TypeError` or `Symfony\Component\Form\Exception\ExceptionInterface` from a 'set' accessor are caught and the exception message added as a form error. This causes the form to fail validation before any changes are applied. However, the exception does not stop the processing of form fields, so this can serve as a sanity check before doing more significant validation (e.g. uniqueness checks).

## Compound fields

Example: we have a checkout form that contains address fields. The fields are added as single form fields, but should all map to a single accessor. The RichModelFormsBundle mentioned in the RFC makes this possible by defining the same `write_property_path` multiple times (from [the documentation](https://github.com/sensiolabs-de/rich-model-forms-bundle/blob/main/docs/mapping.md#mapping-several-form-fields-to-a-single-method)):
```php
$builder
    ->add('address', AddressType::class, [
        'read_property_path' => 'shippingAddress',
        'write_property_path' => 'ship',
    ])
    ->add('trackingNumber', TextType::class, [
        'read_property_path' => 'trackingNumber',
        'write_property_path' => 'ship',
    ])
 ;
```

With closures being used for accessors, this is no longer feasible. It is thus suggested to set your own set accessor:

```php
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        parent::buildForm($builder, $options);

        $builder
            ->add('address', AddressType::class)
            ->add('trackingNumber', TextType::class)
        ;
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        parent::configureOptions($resolver);

        $resolver->setDefault('set', function (Order $order, array $data) { $order->ship($data['address'], $data['trackingNumber']);
    }
```

This use case is also tested in the `testSetUsesAccessorForCompoundFields` test for the access mapper itself.

### CODE REVIEW

**Comment by julienfalque** [comment]
This looks very interesting! This is kind of similar to something I proposed for the serializer (#37419).

> Any `TypeError` or `Symfony\Component\Form\Exception\ExceptionInterface` from a 'set' accessor are caught and the exception message added as a form error.

What do you think about making the classes of caught exceptions configurable somehow? The use case would be the same as in my proposal: rely on invariants validation from Value Objects to validate input instead of duplicating those rules in the Validator mapping. I think we could completely avoid using the Validator in simple scenarios but I'm afraid `TypeError` would not cover all the cases as most invariants violations are reported using e.g. `InvalidArgumentException`. `Symfony\Component\Form\Exception\ExceptionInterface` could be useful but one might prefer not having its domain code depend on it.

**Comment by alcaeus** [comment]
Configurable exception handling is on the to-do list. For the time being I used a pre-existing exception interface. However, users may want to extend or implement a generic DomainException that is then handled. I prefer this expressive usage of an exception interface over another form option.

**Review by Taluu — APPROVED** [review]

File: src/Symfony/Component/Form/Extension/Core/DataMapper/AccessorMapper.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            $this->fallbackMapper->mapFormsToData($forms, $data);
+            return;
+        }
+
+        foreach ($forms as $form) {
+            $config = $form->getConfig();
+
+            // Write-back is disabled if the form is not synchronized (transformation failed),
+            // if the form was not submitted and if the form is disabled (modification not allowed)
+            if (null !== $this->set && $config->getMapped() && $form->isSubmitted() && $form->isSynchronized() && !$form->isDisabled()) {
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Symfony/Component/Form/Extension/Core/DataMapper/AccessorMapper.php:74]: You are correct. I'll update this.

**Comment by ro0NL** [comment]
what generally bugs me, still, is each individually field knowing about mapping rules. Either on the compound form (default option), or per child form (given option). This increases complexity IMHO, as hydration can - and probably will - scatter around the codebase, making it hard to grasp what's going on, on high level.

This is why i ultimately think we should reduce complexity in forms by a lot, letting the user hydrate as nessecary - in valid state, thus more or less

```
succes-handler => fn (FormData $data) {
    $dto->setField($data->getModelData('field'));
    $data->setFieldsAutomatically($dto, ['other-field']);
}
error-handler => fn() {}
```

(FormData to retain access on both view- and model-data)

**Comment by alcaeus** [comment]
> what generally bugs me, still, is each individually field knowing about mapping rules. Either on the compound form (default option), or per child form (given option). This increases complexity IMHO, as hydration can - and probably will - scatter around the codebase, making it hard to grasp what's going on, on high level.

You are right, but that's already the case, it's just hidden from view: the `property_path` option defines the rules on how it's accessed. From there, the actual way it is accessed (direct property access, accessors, reflection, whatever) are hidden from view. All that matters is the property path (which is deduced from the name if not explicitly set). Think of this PR as making these rules more explicit: your form will always have to know how to map its data back to yours, unless you want to do that yourself completely (e.g. in an `apply` method of a DTO). Even in that case, the form still would have to know about how to bind data to your DTO.

What I'm suggesting is that we put this coupled logic together: the form needs to know about the DTO, so it makes sense that this kind of mapping logic is defined in the form. As for the other examples, I always encounter people suggesting building your own data mapper. In my experience, this is quite cumbersome:
```php
public function mapFormsToData(...)
{
    foreach ($forms as $form) {
        switch ($form->getName()) {
            case 'name':
                $data->rename($form->getValue());
                break;

            case 'visible':
                $data->makeVisible();
                break;

            default:
                // Exception?
        }
    }
}
```

This is not very practical, regardless of whether you need to change behaviour for a single field or for multiple fields.

> This is why i ultimately think we should reduce complexity in forms by a lot, letting the user hydrate as nessecary - in valid state, thus more or less
> [...]

Your example requires you to write an apply method that applies form data to your DTO or entity. It would also require you to pass a handler that initialises it, essentially forcing everyone to inline their data mappers. Again, this is not very practical, as many forms will be rather straightforward.

**Comment by ro0NL** [comment]
> > each individually field knowing about mapping rules. 
>
> but that's already the case

true, but it's, IMHO, what causes the form complexity in general. It's over-abstract if you ask me :)

> the form needs to know about the DTO

ultimately i disagree, the user knows about the DTO/subject in relation to a/its form. And it's matter of composition basically, glueing independent concerns.

Note, i wont block the PR (not that i can anyway :P); i think it's a good improvement from the current design pov overall  :+1:

**Comment by alcaeus** [comment]
> Note, i wont block the PR (not that i can anyway :P); i think it's a good improvement from the current design pov overall 👍

Thank you! I'm planning to make some more improvements to forms with respect to DTOs, so your input is always welcome!

**Review by yceruto — CHANGES_REQUESTED** [review]
Thank you for proposing it, this seems a very interesting feature. Although I would use it for single fields only rather than compound ones. For compound forms I'm not sure what would be the advantage compared to `DataMapperInterface`.

**Comment by yceruto** [comment]
> As for the other examples, I always encounter people suggesting building your own data mapper. In my experience, this is quite cumbersome

Please note that this can be done in another way:
```php
public function mapFormsToData(...)
{
    $forms = iterator_to_array($forms);

    $category->rename($forms['name']->getData());

    if ($forms['visible']->getData()) {
        $category->makeVisible();
    }

    // ...
}
```

File: src/Symfony/Component/Form/Extension/Core/DataMapper/AccessorMapper.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                    $form->addError(new FormError($e->getMessage()));
+                    continue;
+                }
+
+                $type = is_object($returnValue) ? get_class($returnValue) : gettype($returnValue);
+
+                if (
+                    (is_scalar($data) && gettype($data) === $type)
+                    || (is_array($data) && is_array($returnValue))
+                    || (is_object($data) && $returnValue instanceof $type)) {
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/DataMapper/AccessorMapper.php:87]: [Immutability] We could also apply the same approach as data mappers but on the closure `&$data` and get rid of extra checks and rules in our side, so it's the user responsability and it works same:
```php
$builder->add('name', null, [
    'set' => function(Category &$category, string $name) { 
        $category = $category->rename($value); 
    },
]);
```

Comment (alcaeus) [reply src/Symfony/Component/Form/Extension/Core/DataMapper/AccessorMapper.php:87]: Interesting idea. Passing the argument by reference also makes it clearer that you are always modifying the data directly, apart from being consistent with the data mapper API.

File: src/Symfony/Component/Form/Extension/Core/Type/AccessorMapperExtension.php [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+class AccessorMapperExtension extends AbstractTypeExtension
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if (!$options['get'] && !$options['set']) {
+            return;
+        }
+
+        if (!$dataMapper = $builder->getDataMapper()) {
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Symfony/Component/Form/Extension/Core/Type/AccessorMapperExtension.php:28]: Ah, yes. I'll add some functional tests to cover this as well.

**Comment by yceruto** [comment]
Hi @alcaeus, I took over your proposal with a different implementation, but same goal ! 
I've taken some ideas from your comments, 
Thank you a lot.

**Comment by fabpot** [comment]
@alcaeus Can you have a look at @yceruto proposal and see if that could replace your PR here? Thank you.

**Comment by alcaeus** [comment]
Closing as the approach by @yceruto makes this a core concept of the forms instead of just tacking it on via an extension.

**Pull request closed without merging**